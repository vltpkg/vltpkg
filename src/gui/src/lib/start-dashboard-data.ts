import type { Action, State } from '@/state/types.ts'

export type RequestDashboardDataOptions = {
  stamp: State['stamp']
}

export type StartDashboardDataOptions = {
  navigate: (route: string) => void
  updateDashboard: Action['updateDashboard']
  updateErrorCause: Action['updateErrorCause']
  stamp: State['stamp']
}

export const requestDashboardData = async ({
  stamp,
}: RequestDashboardDataOptions): Promise<State['dashboard']> => {
  const req = await fetch('./dashboard.json?random=' + stamp)
  const res = (await req.json()) as State['dashboard']
  return res
}

export const startDashboardData = ({
  navigate,
  updateDashboard,
  updateErrorCause,
  stamp,
}: StartDashboardDataOptions) => {
  async function _startDashboard() {
    const dashboardData = await requestDashboardData({
      stamp,
    })
    updateDashboard(dashboardData)
  }

  void _startDashboard().catch((err: unknown) => {
    console.error(err)
    navigate('/error')
    updateErrorCause('Failed to initialize dashboard.')
  })
}
