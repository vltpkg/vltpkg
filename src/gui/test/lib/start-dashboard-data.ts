import { afterEach, beforeEach, expect, test, vi } from 'vitest'
import { startDashboardData } from '@/lib/start-dashboard-data.js'

const stderr = console.error

afterEach(() => {
  console.error = stderr
  vi.clearAllMocks()
})

beforeEach(() => {
  global.fetch = vi.fn(async url => ({
    json: async () => {
      switch (new URL(`file://${url}`).searchParams.get('random')) {
        case 'standard': {
          return {
            buildVersion: '1.0.0',
            cwd: '/path/to/cwd',
            projects: [
              {
                name: 'project-foo',
                readablePath: '~/project-foo',
                path: '/home/user/project-foo',
                manifest: { name: 'project-foo', version: '1.0.0' },
                tools: ['node', 'vlt'],
                mtime: 1730498483044,
              },
              {
                name: 'project-bar',
                readablePath: '~/project-foo',
                path: '/home/user/project-bar',
                manifest: { name: 'project-bar', version: '1.0.0' },
                tools: ['pnpm'],
                mtime: 1730498491029,
              },
            ],
          }
        }
        default: {
          throw Object.assign({ status: 404 }, new Error('Not found'))
        }
      }
    },
  })) as unknown as typeof global.fetch
})

test('standard dashboard response', async () => {
  await new Promise<void>((res, rej) => {
    try {
      startDashboardData({
        updateActiveRoute: vi.fn(),
        updateDashboard: dashboardData => {
          expect(dashboardData).toMatchSnapshot(
            'should return standard dashboard data',
          )
          res()
        },
        updateErrorCause: rej,
        stamp: 'standard',
      })
    } catch (err) {
      rej(err as Error)
    }
  })
})

test('missing dashboard response', async () => {
  console.error = vi.fn()
  await new Promise<void>((res, rej) => {
    try {
      startDashboardData({
        updateActiveRoute: route => {
          expect(route).toBe('/error')
          res()
        },
        updateDashboard: () => {
          throw new Error('Should not have dashboard data')
        },
        updateErrorCause: vi.fn(),
        stamp: 'missing',
      })
    } catch (err) {
      rej(err as Error)
    }
  })
})
