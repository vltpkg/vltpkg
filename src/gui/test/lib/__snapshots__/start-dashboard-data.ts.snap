// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`standard dashboard response > should return standard dashboard data 1`] = `
{
  "buildVersion": "1.0.0",
  "cwd": "/path/to/cwd",
  "projects": [
    {
      "manifest": {
        "name": "project-foo",
        "version": "1.0.0",
      },
      "mtime": 1730498483044,
      "name": "project-foo",
      "path": "/home/user/project-foo",
      "readablePath": "~/project-foo",
      "tools": [
        "node",
        "vlt",
      ],
    },
    {
      "manifest": {
        "name": "project-bar",
        "version": "1.0.0",
      },
      "mtime": 1730498491029,
      "name": "project-bar",
      "path": "/home/user/project-bar",
      "readablePath": "~/project-foo",
      "tools": [
        "pnpm",
      ],
    },
  ],
}
`;
